src/source/raster_tile_source.js

```diff
@@ -238,7 +238,20 @@ class RasterTileSource extends Evented implements Source {
 
     // $FlowFixMe[method-unbinding]
     unloadTile(tile: Tile, callback: Callback<void>) {
-        if (tile.texture) this.map.painter.saveTileTexture(tile.texture);
+        // Cache the tile texture to avoid re-allocating Textures if they'll just be reloaded
+        if (tile.texture && tile.texture instanceof Texture) {
+            // Clean everything else up owned by the tile, but preserve the texture.
+            // Destroy first to prevent racing with the texture cache being popped.
+            tile.destroy(true);
+
+            // Save the texture to the cache
+            if (tile.texture && tile.texture instanceof Texture) {
+                this.map.painter.saveTileTexture(tile.texture);
+            }
+        } else {
+            tile.destroy();
+        }
+
```

Comment:
Just wondering — would it simplify things if we moved this whole logic to `tile.destroy` instead of having a conditional parameter? It might be better for consistency, since the counterpart `getTileTexture` logic of reusing textures is inside `tile.js` (`setTexture` method).

Reply:
I was thinking that too. My thinking against was, if you call destroy() on something you think it should be destroyed. Or at least provide a mechanism which makes sense. I also didn't want to move the calls around too much :)